File: GoLang/src/tsp/solver/solver.go
package solver

import (
	"fmt"
	tsp_types "tsp/types"
)

func calculate_plus_cost(matrix tsp_types.MatrixType, n int) (bool, tsp_types.DataType) {
	cost := tsp_types.DataType(0)
	// every row
	for i := 0; i < n; i++ {
		min_value := tsp_types.POSITIVE_INF
		// looking min value in row
		infinity_count := 0
		for j := 0; j < n; j++ {
			if matrix[i*n+j] == tsp_types.POSITIVE_INF {
				infinity_count++
			} else if matrix[i*n+j] < min_value {
				min_value = matrix[i*n+j]
			}
		}
		// if all elements in row are infinite then return
		if infinity_count == n {
			return false, tsp_types.DataType(0)
		}
		if min_value != 0 {
			// subtract min value from entire row
			for j := 0; j < n; j++ {
				if matrix[i*n+j] != tsp_types.POSITIVE_INF {
					matrix[i*n+j] -= min_value
				}
			}
			// add min value to cost
			cost += min_value
		}
	}
	// every collum
	for i := 0; i < n; i++ {
		min_value := tsp_types.POSITIVE_INF
		// looking min value in collum
		infinity_count := 0
		for j := 0; j < n; j++ {
			if matrix[j*n+i] == tsp_types.POSITIVE_INF {
				infinity_count++
			} else if matrix[j*n+i] < min_value {
				min_value = matrix[j*n+i]
			}
		}
		// if all elements in collum are infinite then return
		if infinity_count == n {
			return false, tsp_types.DataType(0)
		}
		if min_value != 0 {
			// subtract min value from entire column
			for j := 0; j < n; j++ {
				if matrix[j*n+i] != tsp_types.POSITIVE_INF {
					matrix[j*n+i] -= min_value
				}
			}
			// add min value to cost
			cost += min_value
		}
	}
	return true, cost
}

func find_heavier_zero(matrix tsp_types.MatrixType, n int) tsp_types.ZeroInfoType {
	heavier_zero := tsp_types.ZeroInfoType{0, 0, tsp_types.NEGATIVE_INF}
	for i := 0; i < n; i++ {
		// for every zero in row
		for j := 0; j < n; j++ {
			if matrix[i*n+j] == 0 {
				// find min element in row
				weight := tsp_types.DataType(0)
				min_value := tsp_types.POSITIVE_INF
				for z := 0; z < n; z++ {
					if (z != j) && (matrix[i*n+z] < min_value) && (matrix[i*n+z] != tsp_types.POSITIVE_INF) {
						min_value = matrix[i*n+z]
					}
				}
				if min_value != tsp_types.POSITIVE_INF {
					weight += min_value
				}
				// find min element in collum
				min_value = tsp_types.POSITIVE_INF
				for z := 0; z < n; z++ {
					if (z != i) && (matrix[z*n+j] < min_value) && (matrix[z*n+j] != tsp_types.POSITIVE_INF) {
						min_value = matrix[z*n+j]
					}
				}
				if min_value != tsp_types.POSITIVE_INF {
					weight += min_value
				}
				// remember new zero
				if heavier_zero.Weight < weight {
					heavier_zero = tsp_types.ZeroInfoType{i, j, weight}
				}
			}
		}
	}
	return heavier_zero
}

func find_previous_jump(all_jumps []tsp_types.JumpType, current_jump tsp_types.JumpType) (tsp_types.JumpType, bool) {
	for i := 0; i < len(all_jumps); i++ {
		if all_jumps[i].Destination == current_jump.Source {
			return all_jumps[i], true
		}
	}
	return current_jump, false
}

func find_next_jump(all_jumps []tsp_types.JumpType, current_jump tsp_types.JumpType) (tsp_types.JumpType, bool) {
	for i := 0; i < len(all_jumps); i++ {
		if all_jumps[i].Source == current_jump.Destination {
			return all_jumps[i], true
		}
	}
	return current_jump, false
}

func find_last_jump_in_direction(all_jumps []tsp_types.JumpType, current_jump tsp_types.JumpType, direction tsp_types.DirectionType) (r1 tsp_types.JumpType, r2 bool) {
	found_new := false
	last_jump := current_jump
	for {
		switch direction {
		case tsp_types.FORWARD_DIR:
			last_jump, found_new = find_next_jump(all_jumps, last_jump)
		case tsp_types.BACKWARD_DIR:
			last_jump, found_new = find_previous_jump(all_jumps, last_jump)
		default:
			fmt.Println("[find_last_jump_in_direction] Error: No right direction!")
			return tsp_types.JumpType{-1, -1}, false
		}
		if found_new == false {
			break
		}
	}
	if (current_jump.Source != last_jump.Source) && (current_jump.Destination != last_jump.Destination) {
		return last_jump, true
	} else {
		return last_jump, false
	}
}

func forbid_jump_if_needed(matrix *tsp_types.MatrixType, x_mapping []int, y_mapping []int, all_jumps []tsp_types.JumpType, n int) {
	var inf_x, inf_y int
	beginning_of_chain, found_next := find_last_jump_in_direction(all_jumps[1:], all_jumps[0], tsp_types.BACKWARD_DIR)
	end_of_chain, found_prev := find_last_jump_in_direction(all_jumps[1:], all_jumps[0], tsp_types.FORWARD_DIR)
	if !(found_prev && found_next) {
		for i := 0; i < n; i++ {
			if x_mapping[i] == end_of_chain.Destination {
				inf_x = i
			}
			if y_mapping[i] == beginning_of_chain.Source {
				inf_y = i
			}
		}
		(*matrix)[inf_x*n+inf_y] = tsp_types.POSITIVE_INF
	}
}

func generate_sub_task_data(matrix *tsp_types.MatrixType, x_mapping []int, y_mapping []int, all_jumps []tsp_types.JumpType, zero_with_most_weight tsp_types.ZeroInfoType, n int) tsp_types.SubTaskDataType {
	var new_matrix tsp_types.MatrixType = make([]tsp_types.DataType, (n-1)*(n-1))
	new_x_mapping := make([]int, n-1)
	new_y_mapping := make([]int, n-1)
	new_all_jumps := make([]tsp_types.JumpType, 1)

	zero_x := zero_with_most_weight.Row
	zero_y := zero_with_most_weight.Col
	new_jump := tsp_types.JumpType{
		x_mapping[zero_with_most_weight.Row],
		y_mapping[zero_with_most_weight.Col],
	}
	new_all_jumps[0] = new_jump
	for j := range all_jumps {
		new_all_jumps = append(new_all_jumps, all_jumps[j])
	}
	// prepare new x_mapping, y_mapping and matrix.
	i_x := 0
	i_y := 0
	for i := 0; i < n; i++ {
		if i != zero_x {
			new_x_mapping[i_x] = x_mapping[i]
			j_y := 0
			for j := 0; j < n; j++ {
				if j == zero_y {
					continue
				}
				new_matrix[i_x*(n-1)+j_y] = (*matrix)[i*n+j]
				j_y++
			}
		} else {
			i_x--
		}
		if i != zero_y {
			new_y_mapping[i_y] = y_mapping[i]
		} else {
			i_y--
		}
		i_x++
		i_y++
	}
	//fmt.Printf("[generate]** \n")
	forbid_jump_if_needed(&new_matrix, new_x_mapping, new_y_mapping, new_all_jumps, n-1)
	return tsp_types.SubTaskDataType{&new_matrix, new_x_mapping, new_y_mapping, new_all_jumps}
}

func SolveImpl(task tsp_types.TaskType, gl_min_cost *tsp_types.GlobalCostType) (tsp_types.AnswerType, bool) {
	if gl_min_cost.Get() < task.MinCost {
		return tsp_types.ERROR_ANSWER, false
	}
	if (task.Size == 0) || (task.Size == 1) {
		return tsp_types.ERROR_ANSWER, true
	}
	success, additional_cost := calculate_plus_cost(*task.Matrix, task.Size)
	if !success {
		return tsp_types.ERROR_ANSWER, true
	}
	task.CurrCost += additional_cost
	heavier_zero := find_heavier_zero(*task.Matrix, task.Size)
	// if task.Size is 2 end recursion
	if task.Size == 2 {
		next_city := heavier_zero.Col
		previous_city := heavier_zero.Row
		if ((*task.Matrix)[(previous_city^1)*2+(next_city^1)] == 0) &&
			((*task.Matrix)[(previous_city^1)*2+next_city] == tsp_types.POSITIVE_INF) &&
			((*task.Matrix)[previous_city*2+(next_city^1)] == tsp_types.POSITIVE_INF) {
			jumps := make([]tsp_types.JumpType, 2)
			jumps[0] = tsp_types.JumpType{task.RowMapping[heavier_zero.Row], task.ColMapping[heavier_zero.Col]}
			jumps[1] = tsp_types.JumpType{task.RowMapping[heavier_zero.Row^1], task.ColMapping[heavier_zero.Col^1]}
			for j := range task.Jumps {
				jumps = append(jumps, task.Jumps[j])
			}
			return tsp_types.AnswerType{jumps, task.CurrCost}, true
		} else {
			return tsp_types.ERROR_ANSWER, true
		}
	}
	// prepare data for recursive call
	sub_dt := generate_sub_task_data(task.Matrix, task.RowMapping, task.ColMapping, task.Jumps, heavier_zero, task.Size)
	//call this function recursively
	answer, less_global := SolveImpl(tsp_types.TaskType{sub_dt.Matrix, sub_dt.RowMapping, sub_dt.ColMapping, sub_dt.Jumps, task.CurrCost, task.MinCost, task.Size - 1}, gl_min_cost)
	if !less_global {
		return tsp_types.ERROR_ANSWER, false
	}
	//print_answer(answer, "GOT ANSWER")
	final_path := []tsp_types.JumpType{}
	if answer.Cost < task.MinCost {
		final_path = answer.Jumps
		task.MinCost = answer.Cost
	}
	//right_path
	if (task.CurrCost + heavier_zero.Weight) < task.MinCost {
		//correct first one
		(*task.Matrix)[heavier_zero.Row*task.Size+heavier_zero.Col] = tsp_types.POSITIVE_INF
		answer, less_global = SolveImpl(task, gl_min_cost)
		if !less_global {
			return tsp_types.ERROR_ANSWER, false
		}
		if answer.Cost < task.MinCost {
			final_path = answer.Jumps
			task.MinCost = answer.Cost
		}
	}
	// return answer
	if task.MinCost < tsp_types.POSITIVE_INF {
		return tsp_types.AnswerType{final_path, task.MinCost}, true
	} else {
		return tsp_types.ERROR_ANSWER, true
	}
}

func CrusherImpl(task *tsp_types.TaskType) (bool, *tsp_types.TaskType, *tsp_types.TaskType) {
	success, additional_cost := calculate_plus_cost(*task.Matrix, task.Size)
	if success != true {
		return false, nil, nil
	}
	task.CurrCost += additional_cost
	heavier_zero := find_heavier_zero(*task.Matrix, task.Size)
	//prepare data for recursive call
	sub_dt := generate_sub_task_data(task.Matrix, task.RowMapping, task.ColMapping, task.Jumps, heavier_zero, task.Size)
	//call this function recursively
	task1 := tsp_types.TaskType{sub_dt.Matrix, sub_dt.RowMapping, sub_dt.ColMapping, sub_dt.Jumps, task.CurrCost, task.MinCost, task.Size - 1}
	(*task.Matrix)[heavier_zero.Row*task.Size+heavier_zero.Col] = tsp_types.POSITIVE_INF
	return true, &task1, task
}

File: CSharpDev/Solver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TspSolver
{
    public class Solver
    {
        private System.IO.StreamWriter outfile;

        public void print_square_matrix(MatrixType matrix)
        {
            for (int i = 0; i < matrix.Size; ++i)
            {
                for (int j = 0; j < matrix.Size; ++j)
                {
                    if (matrix.Values[i * matrix.Size + j] == Constants.POSITIVE_INF)
                    {
                        outfile.Write(String.Format("{0,5:0}", "INF"));
                    }
                    else
                    {
                        outfile.Write(String.Format("{0,5:0}", matrix.Values[i * matrix.Size + j]));
                    }
                }
                outfile.WriteLine("");
            }
        }

        public Solver()
        {
            outfile = new System.IO.StreamWriter(@"W:\CSharpProjects\TspSolver\TspSolver\res.txt");
        }

        public Tuple<bool, int> calculate_plus_cost(MatrixType matrix)
        {
	        int cost = 0;
	        // every row
	        for (int i = 0; i < matrix.Size; ++i)
            {
                int min_value = Constants.POSITIVE_INF;
                // looking min value in row
                int infinity_count = 0;
		        for (int j = 0; j < matrix.Size; ++j)
                {
			        if (matrix.Values[i * matrix.Size + j] == Constants.POSITIVE_INF)
                    {
                        ++infinity_count;
			        }
                    else if (matrix.Values[i * matrix.Size + j] < min_value)
                    {
                        min_value = matrix.Values[i * matrix.Size + j];
                    }
		        }
		        // if all elements in row are infinite then return
		        if (infinity_count == matrix.Size)
                {
			        return Tuple.Create(false, 0);
		        }
		        if (min_value != 0)
                {
			        // subtract min value from entire row
			        for (int j = 0; j < matrix.Size; ++j)
                    {
				        if (matrix.Values[i * matrix.Size + j] != Constants.POSITIVE_INF)
                        {
                            matrix.Values[i * matrix.Size + j] -= min_value;
				        }
			        }
                    // add min value to cost
                    cost += min_value;
		        }
	        }
	        // every collum
	        for (int i = 0; i < matrix.Size; ++i)
            {
                int min_value = Constants.POSITIVE_INF;
                // looking min value in collum
                int infinity_count = 0;
		        for (int j = 0; j < matrix.Size; ++j)
                {
			        if (matrix.Values[j * matrix.Size + i] == Constants.POSITIVE_INF)
                    {
                        ++infinity_count;
			        }
                    else if (matrix.Values[j * matrix.Size + i] < min_value)
                    {
                        min_value = matrix.Values[j * matrix.Size + i];
			        }
		        }
		        // if all elements in collum are infinite then return
		        if (infinity_count == matrix.Size)
                {
			        return Tuple.Create(false, 0);
                }
		        if (min_value != 0)
                {
			        // subtract min value from entire column
			        for (int j = 0; j < matrix.Size; ++j)
                    {
				        if (matrix.Values[j * matrix.Size + i] != Constants.POSITIVE_INF)
                        {
                            matrix.Values[j * matrix.Size + i] -= min_value;
				        }
			        }
                    // add min value to cost
                    cost += min_value;
		        }
	        }
            return Tuple.Create(true, cost);
        }

        public ZeroInfo find_heaviest_zero(MatrixType matrix)
        {
            ZeroInfo heaviest_zero = new ZeroInfo(0, 0, Constants.NEGATIVE_INF);
	        for (int i = 0; i < matrix.Size; ++i)
            {
		        // for every zero in row
		        for (int j = 0; j < matrix.Size; ++j)
                {
			        if (matrix.Values[i * matrix.Size + j] == 0)
                    {
                        // find min element in row
                        int weight = 0;
				        int min_value = Constants.POSITIVE_INF;
				        for (int z = 0; z < matrix.Size; ++z)
                        {
					        if ((z != j) && (matrix.Values[i * matrix.Size + z] < min_value) &&
                                (matrix.Values[i * matrix.Size + z] != Constants.POSITIVE_INF))
                            {
                                min_value = matrix.Values[i * matrix.Size + z];
					        }
				        }
				        if (min_value != Constants.POSITIVE_INF)
                        {
                            weight += min_value;
				        }
                        // find min element in collum
                        min_value = Constants.POSITIVE_INF;
				        for (int z = 0; z < matrix.Size; ++z)
                        {
					        if ((z != i) && (matrix.Values[z * matrix.Size + j] < min_value) &&
                                (matrix.Values[z * matrix.Size + j] != Constants.POSITIVE_INF))
                            {
                                min_value = matrix.Values[z * matrix.Size + j];
					        }
				        }
				        if (min_value != Constants.POSITIVE_INF)
                        {
                            weight += min_value;
				        }
				        // remember new zero
				        if (heaviest_zero.Weight < weight)
                        {
                            heaviest_zero = new ZeroInfo(i, j, weight);
				        }
			        }
		        }
	        }
            return heaviest_zero;
        }

        public Tuple<Jump, bool> find_previous_jump(Jump[] all_jumps, Jump current_jump)
        {
	        for (int i = 0; i < all_jumps.Length; ++i)
            {
		        if (all_jumps[i].Destination == current_jump.Source)
                {
			        return Tuple.Create(all_jumps[i], true);
		        }
	        }
            return Tuple.Create(current_jump, false);
        }

        public Tuple<Jump, bool> find_next_jump(Jump[] all_jumps, Jump current_jump)
        {
	        for (int i = 0; i < all_jumps.Length; ++i)
            {
		        if (all_jumps[i].Source == current_jump.Destination)
                {
			        return Tuple.Create(all_jumps[i], true);
                }
	        }
	        return Tuple.Create(current_jump, false);
        }

        public Tuple<Jump, bool> find_last_jump_in_direction(Jump[] all_jumps, Jump current_jump, Direction direction)
        {
            bool found_new = false;
            Jump last_jump = current_jump;
            while (true)
            {
                switch (direction)
                {
                    case Direction.FORWARD_DIR:
                        var tuple1 = find_next_jump(all_jumps, last_jump);
                        last_jump = tuple1.Item1;
                        found_new = tuple1.Item2;
                        break;
                    case Direction.BACKWARD_DIR:
                        var tuple2 = find_previous_jump(all_jumps, last_jump);
                        last_jump = tuple2.Item1;
                        found_new = tuple2.Item2;
                        break;
                    default:
                        Console.WriteLine("[find_last_jump_in_direction] Error: No right direction!");
                        return Tuple.Create(new Jump(-1, -1), false);
                }
                if (found_new == false)
                {
                    break;
                }
            }
            if ((current_jump.Source != last_jump.Source) && (current_jump.Destination != last_jump.Destination))
            {
                return Tuple.Create(last_jump, true);
            }
            else
            {
                return Tuple.Create(last_jump, false);
            }
        }

        public void forbid_jump_if_needed(MatrixType matrix, int[] row_mapping, int[] col_mapping, Jump[] all_jumps)
        {
            int inf_x = 0, inf_y = 0;
            Jump[] jumps = new Jump[all_jumps.Length - 1];
            for (int i = 1; i < all_jumps.Length; ++i)
            {
                jumps[i - 1] = all_jumps[i];
            }
            var tuple1 = find_last_jump_in_direction(jumps, all_jumps[0], Direction.BACKWARD_DIR);
            var tuple2 = find_last_jump_in_direction(jumps, all_jumps[0], Direction.FORWARD_DIR);
            Jump beginning_of_chain = tuple1.Item1;
            bool found_next = tuple1.Item2;
            Jump end_of_chain = tuple2.Item1;
            bool found_prev = tuple2.Item2;
            outfile.WriteLine("SolveImpl 3.1.1 {" +
                beginning_of_chain.Source.ToString() + ";" +
                beginning_of_chain.Destination.ToString() + "} " +
                found_next.ToString() + " {" +
                end_of_chain.Source.ToString() + ";" +
                end_of_chain.Destination.ToString() + "} " +
                found_prev.ToString());
            if (!(found_prev && found_next))
            {
                outfile.WriteLine("cond 1");
                for (int i = 0; i < matrix.Size; ++i)
                {
                    if (row_mapping[i] == end_of_chain.Destination)
                    {
                        inf_x = i;
                    }
                    if (col_mapping[i] == beginning_of_chain.Source)
                    {
                        inf_y = i;
                    }
                }
                outfile.WriteLine("row = " + inf_x.ToString() + ", col = " + inf_y.ToString());
                matrix.Values[inf_x * matrix.Size + inf_y] = Constants.POSITIVE_INF;
            }
            else
            {
                outfile.WriteLine("cond 2");
            }
        }

        public SubTaskData generate_sub_task_data(MatrixType matrix, int[] row_mapping, int[] col_mapping, Jump[] all_jumps, ZeroInfo heaviest_zero)
        {
            MatrixType new_matrix = new MatrixType(matrix.Size - 1);
            int[] new_row_mapping = new int[matrix.Size - 1];
            int[] new_col_mapping = new int[matrix.Size - 1];
            Jump[] new_all_jumps = new Jump[ (all_jumps != null) ? all_jumps.Length + 1 : 1];
            int zero_row = heaviest_zero.Row;
            int zero_col = heaviest_zero.Col;
	        new_all_jumps[0] = new Jump(row_mapping[zero_row], col_mapping[zero_col]);
            if (all_jumps != null)
            {
                for (int j = 0; j < all_jumps.Length; ++j)
                {
                    new_all_jumps[j + 1] = all_jumps[j];
                }
            }
            // prepare new x_mapping, y_mapping and matrix.
            int i_row = 0, i_col = 0;
	        for (int i = 0; i < matrix.Size; ++i)
            {
		        if (i != zero_row)
                {
                    new_row_mapping[i_row] = row_mapping[i];
                    int j_col = 0;
			        for (int j = 0; j < matrix.Size; ++j)
                    {
				        if (j == zero_col)
                        {
                            continue;
				        }
                        new_matrix.Values[i_row * (matrix.Size - 1) + j_col] = matrix.Values[i * matrix.Size + j];
                        ++j_col;
			        }
		        }
                else
                {
                    --i_row;
		        }
		        if (i != zero_col)
                {
                    new_col_mapping[i_col] = col_mapping[i];
		        }
                else
                {
                    --i_col;
		        }
                ++i_row;
                ++i_col;
	        }
            //fmt.Printf("[generate]** \n")
            //outfile.WriteLine("SolveImpl 3.1");
            //print_square_matrix(new_matrix);
            forbid_jump_if_needed(new_matrix, new_row_mapping, new_col_mapping, new_all_jumps);
            outfile.WriteLine("SolveImpl 3.2");
            print_square_matrix(new_matrix);
            return new SubTaskData(new_matrix, new_row_mapping, new_col_mapping, new_all_jumps);
        }

        public Answer SolveImpl(Task task)
        {
	        if ((task.Matrix.Size == 0) || (task.Matrix.Size == 1)) {
		        return Constants.ERROR_ANSWER;
            }
            outfile.WriteLine("SolveImpl 1");
            print_square_matrix(task.Matrix);
            var tuple1 = calculate_plus_cost(task.Matrix);
            bool success = tuple1.Item1;
            int additional_cost = tuple1.Item2;
            if (!success)
            {
                return Constants.ERROR_ANSWER;
            }
            task.CurrCost += additional_cost;
            ZeroInfo heaviest_zero = find_heaviest_zero(task.Matrix);
            outfile.WriteLine("SolveImpl 2");
            print_square_matrix(task.Matrix);
            // if task.Size is 2 end recursion
            if (task.Matrix.Size == 2)
            {
                int next_city = heaviest_zero.Col;
                int previous_city = heaviest_zero.Row;
		        if ((task.Matrix.Values[(previous_city ^ 1)*2+(next_city^1)] == 0) &&
			        (task.Matrix.Values[(previous_city ^ 1)*2+next_city] == Constants.POSITIVE_INF) &&
			        (task.Matrix.Values[previous_city*2+(next_city^1)] == Constants.POSITIVE_INF))
                {
                    Jump[] jumps = new Jump[task.Jumps.Length+2];
                    jumps[0] = new Jump(task.RowMapping[heaviest_zero.Row], task.ColMapping[heaviest_zero.Col]);
                    jumps[1] = new Jump(task.RowMapping[heaviest_zero.Row ^ 1], task.ColMapping[heaviest_zero.Col ^ 1]);
			        for (int j = 0; j < task.Jumps.Length; ++j)
                    {

                        jumps[j + 2] = task.Jumps[j];
			        }
                    return new Answer(jumps, task.CurrCost);
		        } else {
                    return Constants.ERROR_ANSWER;
                }
	        }
            // prepare data for recursive call
            outfile.WriteLine("SolveImpl 3");
            print_square_matrix(task.Matrix);
            SubTaskData sub_dt = generate_sub_task_data(task.Matrix, task.RowMapping, task.ColMapping, task.Jumps, heaviest_zero);
            //outfile.WriteLine("SolveImpl 3.5");
            //print_square_matrix(sub_dt.Matrix);
            //call this function recursively
            Answer answer = SolveImpl(new Task(sub_dt.Matrix, sub_dt.RowMapping, sub_dt.ColMapping, sub_dt.Jumps, task.CurrCost, task.MinCost));
            //print_answer(answer, "GOT ANSWER")
            Jump[] final_path = null;
	        if (answer.Cost < task.MinCost)
            {
                final_path = answer.Jumps;
                task.MinCost = answer.Cost;
	        }
            outfile.WriteLine("SolveImpl 4");
            print_square_matrix(task.Matrix);
            //right_path
            if ((task.CurrCost + heaviest_zero.Weight) < task.MinCost)
            {
                //correct first one
                task.Matrix.Values[heaviest_zero.Row * task.Matrix.Size + heaviest_zero.Col] = Constants.POSITIVE_INF;
                answer = SolveImpl(task);
		        if (answer.Cost < task.MinCost)
                {
                    final_path = answer.Jumps;
                    task.MinCost = answer.Cost;
		        }
	        }
            outfile.WriteLine("SolveImpl 5");
            // return answer
            if (task.MinCost < Constants.POSITIVE_INF)
            {
                return new Answer(final_path, task.MinCost);
	        } else {
		        return Constants.ERROR_ANSWER;
	        }
        }

        Tuple<bool, Task, Task>  CrusherImpl(Task task)
        {
            var tuple1 = calculate_plus_cost(task.Matrix);
            bool success = tuple1.Item1;
            int additional_cost = tuple1.Item2;
            if (!success) 
            {
                return Tuple.Create(false, new Task(), new Task());
            }
            task.CurrCost += additional_cost;
            ZeroInfo heaviest_zero = find_heaviest_zero(task.Matrix);
            //prepare data for recursive call
            SubTaskData sub_dt = generate_sub_task_data(task.Matrix, task.RowMapping, task.ColMapping, task.Jumps, heaviest_zero);
            //call this function recursively
            Task task1 = new Task(sub_dt.Matrix, sub_dt.RowMapping, sub_dt.ColMapping, sub_dt.Jumps, task.CurrCost, task.MinCost);
            task.Matrix.Values[heaviest_zero.Row * task.Matrix.Size + heaviest_zero.Col] = Constants.POSITIVE_INF;
            return Tuple.Create(true, task1, task);
        }
    }
}
